project_wip_material/task_picking_smart_button.py

```diff
@@ -35,14 +55,52 @@ def open_consumption_picking_view_from_task(self):
 
         odoo/addons/sale_stock/models/sale_order.py
         """
-        action = self.env.ref('stock.action_picking_tree_all').read()[0]
+        return self._open_stock_pickings_view_from_task(self.consumption_picking_ids)
 
-        if self.consumption_picking_count > 1:
-            action['domain'] = [('id', 'in', self.consumption_picking_ids.ids)]
 
-        else:
-            picking_form_view = self.env.ref('stock.view_picking_form')
-            action['views'] = [(picking_form_view.id, 'form')]
-            action['res_id'] = self.consumption_picking_ids.id
+def _is_preparation_return_picking(picking: 'StockPicking'):
+    """Return whether the given stock picking is a return picking.
 
-        return action
+    This function allows to partition `Preparation Pickings` from
+    `Preparation Return Pickings`.
+
+    The strategy is to check whether the picking type is the picking
+    type defined as the preparation return type on the warehouse.
+    """
+    picking_type = picking.picking_type_id
+    return picking_type == picking_type.warehouse_id.consu_prep_return_type_id
+
+
+class TaskWithPreparationPickingSmartButton(models.Model):
+    """Add the smart button to view the preparation pickings."""
+
+    _inherit = 'project.task'
+
+    preparation_picking_count = fields.Integer(compute='_compute_preparation_pickings')
+    preparation_picking_ids = fields.One2many(
+        'stock.picking', string='Preparation Pickings',
+        compute='_compute_preparation_pickings')
+    preparation_return_picking_count = fields.Integer(compute='_compute_preparation_pickings')
+    preparation_return_picking_ids = fields.One2many(
+        'stock.picking', string='Preparation Return Pickings',
+        compute='_compute_preparation_pickings')
+
+    def _compute_preparation_pickings(self):
+        tasks_with_procurement_group = self.filtered(lambda t: t.procurement_group_id)
+        for task in tasks_with_procurement_group:
+            pickings = self.env['stock.picking'].search([
+                ('group_id', '=', task.procurement_group_id.id),
+                ('picking_type_code', '=', 'internal'),
+            ])
+            prep_picking = pickings.filtered(lambda p: not _is_preparation_return_picking(p))
+            return_picking = pickings.filtered(lambda p: _is_preparation_return_picking(p))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Lambda may not be necessary](https://app.codacy.com/app/numigi/odoo-project-addons/pullRequest?prid=4160521)